Fractional reference bin indices for correcting TOF axis

darominski opened the pull request:
**Description of work.**

This PR implements a modification of the direct geometry workflow to improve the precision of the elastic peak calibration. Previously, only the elastic peak time channel was communicated with the CorrectTOFAxis algorithm, which limits the precision of the calibration to the time channel with. In result, the elastic peak centres were observed to not fall at 0 energy exchange but around 0.1 meV away from it.

A solution proposed in this PR adds a possiblity to specify a fractional offset that allows to profit from the full information about the peak centre position obtained from elastic peak fitting. The result was initially validated with Stephane Rols and Bjorn Fak, and the proposed solution was found superior to the one currently available in the code.

**To test:**

<!-- Instructions for testing. -->

1. Code review, tests should be passing.
2. Get the attached file and run the following script in the workbench:

```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import directtools as dt

elp = CreateSingleValuedWorkspace(DataValue=103.42)

DirectILLCollectData( 
        Run = '026079',
        IncidentEnergyCalibration='Energy Calibration ON',
        ElasticChannelWorkspace=elp,  
        OutputWorkspace = 'test'
)
DirectILLReduction(
            InputWorkspace='test',
            OutputWorkspace="out"
)
Q= 2.85
dQ= 0.35 # this means Q in [2.5,3.2]
plot= "out"
fig, axes, cuts = dt.plotconstQ(plot,Q,dQ)
```
The maximum of the plotted distribution should be now very close to 0.0, while when the data value of the ELP is changed to the nearest integer, the peak would fall at about 0.1 meV.

Fixes #33222.
[026079.nxs.gz](https://github.com/mantidproject/mantid/files/7830083/026079.nxs.gz)

Release note added.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

darominski committed:
Changed type of ElasticBinIndex property from int to double.

The floating-point value allows to use fractional bin sizes for higher
degree of control in correcting the TOF axis, e.g. where the integer index
does not perfectly reproduce the fitted elastic peak position.

darominski committed:
CorrectTOFAxis: unit test for floating-point bin index

darominski committed:
[DirectILL] Changed numpy import to np alias

darominski committed:
[DirectILL] ElasticBinIndex for CorrectTOFAxis is now a float

The floating-point value for the index rounded to the second significant digit is now forwarded to
the CorrectTOFAxis ElasticBinIndex property for more precise estimate of the elastic peak centre position
coming from fitting. System tests and reference data updated.

darominski committed:
[DirectILL] updated docs with floating-point elastic channel index

darominski committed:
[DirectILL] Release notes for float elastic bin index

darominski committed:
[DirectILL] Moved float elastic bin index precision to IPFs,

and ensured consistent binning when applying self-shielding correction via RebinToWorkspace.
The default number of significant digits is set to 3 for all supported instruments.
Doc and system tests are adapted to the new precision of the bin index.

darominski committed:
[DirectILL] Release note for the use of matching binning

gui-co commented on Framework/Algorithms/src/CorrectTOFAxis.cpp:
```diff
@@ -338,13 +339,18 @@ void CorrectTOFAxis::useReferenceWorkspace(const API::MatrixWorkspace_sptr &outp
+  auto l2 = 0.0;
+  auto epp = 0.0;
+  auto const elasticBinIndexOffset =
+      static_cast<double>(getProperty(PropertyNames::ELASTIC_BIN_INDEX)) - static_cast<int>(m_elasticBinIndex);
```
Here, `m_elasticBinIndex` is a conversion in workspace index of the integer part of the property `PropertyNames::ELASTIC_BIN_INDEX`. Can you still do this operation?

gui-co reviewed (commented):
The test your are providing is working properly. I added few comments in the code.

darominski commented on Framework/Algorithms/src/CorrectTOFAxis.cpp:
```diff
@@ -338,13 +339,18 @@ void CorrectTOFAxis::useReferenceWorkspace(const API::MatrixWorkspace_sptr &outp
+  auto const elasticBinIndexOffset =
+      static_cast<double>(getProperty(PropertyNames::ELASTIC_BIN_INDEX)) - static_cast<int>(m_elasticBinIndex);
+  auto eppOffset = 0.0;
   g_log.information() << "EPP: " << epp << ".\n";
```
You are absolutely correct. This has been present in the code before this PR. I am going to move this log print after the epp is set.

darominski committed:
Move EPP log information to AFTER it is properly set

darominski committed:
CorrectTOXAxis: EP-bin offset calculation now uses floor operation

Previously, it used a difference between the double value of the provided
index and the integer m_elasticBinIndex value, which can be off as it is
recalculated to the workspace index.

darominski commented:
> The test your are providing is working properly. I added few comments in the code.

I have responded to your comments and taken them into account. Could you have another look at this?

gui-co reviewed (approved):
Thanks for the changes.
